File: 2020/3.kt
// Advent of code 2020 day 3
import java.io.File

fun count_trees(move_right: Int, move_down: Int, lines: List<String>): Int {
  var count: Int = 0
  var x: Int = 0
  var y: Int = 0
  var tree: Char = '#'
  val height: Int = lines.size
  val width: Int = lines[0].length
  while (y < height) {
    if (lines[y][x] == tree) {
      count += 1
    }
    x += move_right
    x = x % width
    y += move_down
  }

  return count
}

fun main() {
  val inputfile = "input3.txt"
  val lines: List<String> = File(inputfile).readLines()

  val slopes = listOf(Pair(1, 1), Pair(3, 1), Pair(5, 1), Pair(7, 1), Pair(1, 2))
  var product: Long = 1
  for (slope in slopes) {
    val count = count_trees(slope.first, slope.second, lines)
    product *= count
  }
  println(product)
}

File: 2020/3.py
#!/usr/bin/env python

inputfile = "input3.txt"
with open(inputfile, "r") as f:
  lines = [x.strip() for x in f.readlines()]

# Treating the string as an array, lovely.
"""
lines = [
  "..##.......",
  "#...#...#..",
  ".#....#..#.",
  "..#.#...#.#",
  ".#...##..#.",
  "..#.##.....",
  ".#.#.#....#",
  ".#........#",
  "#.##...#...",
  "#...##....#",
  ".#..#...#.#",
]
"""

slopes = [(1,1), (3,1), (5,1), (7,1), (1,2)]  # (y, x)!
counts = []

for slope in slopes:
  y = 0
  x = 0
  move_right = slope[0]
  move_down = slope[1]
  count = 0
  while y < len(lines):
    if lines[y][x] == "#":
      count += 1
    x += move_right
    x = x % len(lines[0])
    y += move_down

  print(count)
  counts.append(count)

product = 1
for count in counts:
  product *= count

print ("product:", product)
